### PR TITLE
[zwave-js] proposed lighting changes

### PR DESCRIPTION
**Pull request by andrewfg**
As discussed in the main thread.

@lsiepel the branch on my repository is fully synchronized with OH main (as of today) whereas apparently the branch on your repository is not so. Therefore so you will see in this PR all of the commits on OH main that are not yet synchronized into your branch.

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

### CODE REVIEW

**Comment by lsiepel** [comment]
Updated all my branches and i'm done with the fixes for now. So you can safely update this PR.

**Comment by andrewfg** [comment]
PS let me add a JUnit test based on my light as well..

**Comment by lsiepel** [comment]
Thanks, had to wrap my head around the nodehandler changes, but i see what you did and there are some very good changes.

The `cachedColor` is now applied at Node level. One node might contain multiple HSBType channels (multiple endpoints or property's) that would be problematic. 
A solution could be to have a KeyValue list with key => channel uid / value =?> cachedColor.

But i'm wondering if we can come up with a better solution as i prefer not to have those details in the node handler. Can't we get the actual color values real-time from that channel? That would prevent the caching stuff.

**Comment by andrewfg** [comment]
@lsiepel the status is as follows:

- [ ] color hsb commands working (also command dimmer channels)
- [ ] dimmer commands working
- [ ] color temperature commands working
- [ ] color hsb updates working (also update dimmer channels)
- [ ] dimmer updates working (also update color hsb channels)
- [ ] color temperature updates working (also update color hsb channels)

**Comment by andrewfg** [comment]
Do you know of such a device? So I can buy one for testing..

**Comment by andrewfg** [comment]
Also .. apropos polling vs caching .. there seems to be no method for driver.getNode(index) so I guess we would have to use driver.getnodeS() and parse out what we need. However I think even that call is asynchronous, so we would anyway have to either a) block on the response, or b) use a local cache value.

File: bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java [hunk 1]
```diff
@@ -152,18 +161,66 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+                if (supportsWarmWhite) {
+                    colorMap.put(WARM_WHITE, 0);
+                }
+                zwaveCommand.value = colorMap;
+            }
+            if (isColorChannelCommand) {
+                // schedule brightness command(s) for dimmer channel(s)
+                PercentType brightness = hsbTypeCommand.getBrightness();
+                thing.getChannels().stream().filter(c -> CoreItemFactory.DIMMER.equals(c.getAcceptedItemType()))
+                        .forEach(c -> scheduler.submit(() -> handleCommand(c.getUID(), brightness)));
```
Comment (lsiepel) [inline bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java:222]: Have two question sabou this:
1. A node can have multiple dimmer channels where not all of them are brightness
2. A command can lead up to two commands (HS + B) causing some possible flickering of a lightsource. Have you looked into the API to combine commands to the zwave server?

**Comment by lsiepel** [comment]
> Also .. apropos polling vs caching .. there seems to be no method for driver.getNode(index) so I guess we would have to use driver.getnodeS() and parse out what we need. However I think even that call is asynchronous, so we would anyway have to either a) block on the response, or b) use a local cache value.

Most difficult will be to determine the channel/endpoint. If you know wich one, why don't you ask the framework for the value instead of polling the externel server? 

FYI, look at the `NodeGetValueCommand` in the binding. Based on https://github.com/zwave-js/zwave-js-server?tab=readme-ov-file : 

```javascript

interface {
  messageId: string;
  command: "node.get_value";
  nodeId: number;
  valueId: {
    commandClass: CommandClasses;
    endpoint?: number;
    property: string | number;
    propertyKey?: string | number;
  };
}

```

File: bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java [hunk 2]
```diff
@@ -316,13 +364,24 @@ public boolean onNodeStateChanged(Event event) {
...
+                        }
+                        if (CoreItemFactory.DIMMER.equals(channel.getAcceptedItemType())
+                                && state instanceof PercentType brightness) {
+                            cachedColor = new HSBType(cachedColor.getHue(), cachedColor.getSaturation(), brightness);
+                            thing.getChannels().stream()
+                                    .filter(c -> CoreItemFactory.COLOR.equals(c.getAcceptedItemType()))
+                                    .forEach(c -> updateState(c.getUID(), cachedColor));
+                        }
                     }
+                    updateState(metadata.id, state);
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java:384]: As I mentioned on the other PR I cannot see any reason why the updateState method would throw an exception. It is not described in the OH JavaDoc..

File: bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java (hunk 1 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java:222]: Double checked, combining into 1 message or sending a batch of messages is not supported.

File: bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.zwavejs/src/main/java/org/openhab/binding/zwavejs/internal/handler/ZwaveJSNodeHandler.java:384]: And i responded there to leave it as is, to find it changed here, not directly related to the color suggestion :-)

**Comment by andrewfg** [comment]
The map will need to hold the other colors params too. And the stream actions will need to filter on the respective endpoint as well. I will do it tomorrow.

As mentioned there still is an issue sending color temp commands. Seemingly not yet solved by the propertyKey fix. But maybe I need to recreate the things for the respective chsnnrl config params to be included.

PS are the channel config params read only? It looks like OH configure channel UI is able to change them; which could lead to unforseen problems.

**Comment by andrewfg** [comment]
> don’t expect users to do random weird changes there.

:-/

**Comment by lsiepel** [comment]
> to be the following JSON. Now my understanding is that if there is a warmWhite or coldWhite value present then the color derives NOT from the RGB but from the warmWhite/coldWhite values. => So my question is : Did you actually capture this value in real life (in which case I need to fix the code), or is it a simulation value (

This is straight from the store of my production system, so real-life data.

**Comment by andrewfg** [comment]
> straight from the store of my production system, so real-life data.

Ok. That edge case is now handled correctly as well. So the Junit test does now pass.

**Comment by lsiepel** [comment]
> ing might

I would not expect it, but it is theoretically possible. As it affects the tag and not the functionality of the device itself, i find it acceptable. The current base is very solid for the tagging, i would also expect some more refinements to the tagging logic when the binding gets more users.

**Comment by lsiepel** [comment]
All set for a merge?

**Comment by andrewfg** [comment]
> All set for a merge? From my side all looks good.

Yes. But please note that I had added a short circuit in `handleCommand()` to handle Refresh commands .. with a TODO on it..

**Comment by lsiepel** [comment]
> still have a minor niggle: The light supports two color temperature channels for warm and cold white. Both are Number channels with range 0..255. So to manipulate the CT it requires sending commands to increase (say) the warm value, and decrease (say) the cold value. Ideally I would like to merge those channels into a single Dimmer channel that can be linked to an Item of type `system.color-temperature` resp. `system.color-temperature-abs` so that we can use the UI ColorTemperaturePicker control on it. However that would require some a

AFAICT these two channels are named consistently. So they can be pinpointed, only the logic might be a little bit harder to join them and seperate them into seperate commands to zwave js server.

**Comment by andrewfg** [comment]
> they can be pinpointed, only the logic might be a little bit harder to join them and seperate them into seperate commands to zwave js server.

Shall I try it?

**Comment by lsiepel** [comment]
Not sure how simple it would be, but what about creating one new with the combined functionality. The `old` channels could be marked advanced. or completeley removed. 
wdyt about the same for the individual RGB channels, as there is a single hsb channel?

**Comment by andrewfg** [comment]
We could make the individual component channels advanced. The individual RGBs are already combined in the HSB and the individual CTs would similarly be combined in the uet to be defined new channel.
